### PR TITLE
Better crosspost loading

### PR DESCRIPTION
**Pull request by oetherington**
This PR makes several big improvements to the way crossposts are handled on the front-end.

* We now have proper SSR for crossposts, and separate Apollo clients with caches for local posts/foreign posts on both the server and client
* Crosspost contents are shown in recent discussions and posts list hoverovers (some other places such as sunshine sidebar hoverovers are still left to fix, but the infrastructure in this PR will make that easier to do in a more generic way)
* Crossposted link post links now work instead of just saying "Invalid URL"

I'm sure there are still architectural improvements that can be made here in a future PR, but I'm eager to get this merged ASAP since it fixes several user facing bugs.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203081247965141) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/server/vulcan-lib/apollo-ssr/apolloClient.ts [hunk 1]
```diff
@@ -2,24 +2,33 @@ import { ApolloClient, ApolloLink, InMemoryCache } from '@apollo/client';
 import { SchemaLink } from '@apollo/client/link/schema';
```
Comment (jpaddison3) [inline packages/lesswrong/server/vulcan-lib/apollo-ssr/apolloClient.ts:2]: Now unused?

File: packages/lesswrong/components/hooks/useForeignApolloClient.tsx [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { createContext, useContext } from "react";
+import type { ApolloClient, NormalizedCacheObject } from '@apollo/client';
+
+export const ForeignApolloClientContext = createContext<ApolloClient<NormalizedCacheObject> | undefined>(undefined);
+
+export const useForeignApolloClient = () => useContext(ForeignApolloClientContext);
```
Comment (jpaddison3) [inline packages/lesswrong/components/hooks/useForeignApolloClient.tsx:6]: Flagging that I don't know if this should be done in a context or not. I'll think more about it talking with you.

**Review by jpaddison3 — COMMENTED** [review]
Let's chat about this, but I think this is good. Nice work, there's some tricky things in here I think you've done well.

File: packages/lesswrong/components/hooks/useForeignCrosspost.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+};
+
+export const isPostWithForeignId =
+  <T extends {fmCrosspost: PostsList["fmCrosspost"]}>(post: T): post is T & PostWithForeignId =>
+    !!post.fmCrosspost &&
+    !!post.fmCrosspost.isCrosspost &&
+    typeof post.fmCrosspost.hostedHere === "boolean" &&
+    !!post.fmCrosspost.foreignPostId;
+
+export const useForeignCrosspost = <Post extends PostWithForeignId, FragmentTypeName extends keyof FragmentTypes>(
```
Comment (jpaddison3) [inline packages/lesswrong/components/hooks/useForeignCrosspost.ts:20]: Docstring

File: packages/lesswrong/components/hooks/useForeignCrosspost.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+): {
+  loading: boolean,
+  error?: ApolloError,
+  localPost: Post,
+  foreignPost?: FragmentTypes[FragmentTypeName],
+  combinedPost?: Post & FragmentTypes[FragmentTypeName],
+} => {
+  if (!post.fmCrosspost.foreignPostId) {
+    throw new Error("Crosspost has not been created yet");
+  }
```
[the remark this answers is not in the document]
Comment (jpaddison3) [reply packages/lesswrong/components/hooks/useForeignCrosspost.ts:32]: Cool, we're safe. But we should document.

File: packages/lesswrong/lib/apollo/links.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import type { GraphQLSchema, SourceLocation } from "graphql";
```
[the remark this answers is not in the document]
Comment (jpaddison3) [reply packages/lesswrong/lib/apollo/links.ts:1]: Docstrings on the Links, and top of file link to docs

**Review by jpaddison3 — APPROVED** [review]
ASTLAC, great work

File: packages/lesswrong/components/posts/PostsHighlight.tsx [hunk 6]
```diff
@@ -53,10 +71,80 @@ const PostsHighlight = ({post, maxLengthWords, forceSeeMore=false, classes}: {
...
+
+const ForeignPostsHighlight = ({post, maxLengthWords, forceSeeMore=false, classes}: {
+  post: PostsList & PostWithForeignId,
+  maxLengthWords: number,
+  forceSeeMore?: boolean,
+  classes: ClassesType,
+}) => {
+  const {loading, error, combinedPost} = useForeignCrosspost(post, foreignFetchProps);
+  if (error) {
+    throw error;
```
Comment (b0b3rt) [inline packages/lesswrong/components/posts/PostsHighlight.tsx:86]: Does this get caught more locally than "the top-level app layer"?  I think in most places we get an error from a graphql query we don't throw but render some sort of minimal error component instead.

**Pull request merged**